File: Hamming/python/HammingR.py
def calculate_parity_bits(m):
    positions = []
    acu = 0
    i = 0
    while acu < m:
        acu += 2 ** i
        i += 1
    for j in range(i):
        positions.append(2 ** j)
    return positions


def from_binary_to_decimal(binary):
    decimal = 0
    n = len(binary)
    for i in range(n):
        decimal += int(binary[i]) * 2 ** (n - i - 1)
    return decimal


def is_in_array(array, value):
    for i in range(len(array)):
        if value == array[i]:
            return True
    return False


def even_parity(bits):
    count = 0
    for bit in bits:
        if bit == 1:
            count += 1
    return '0' if count % 2 == 0 else '1'


def odd_parity(bits):
    count = 0
    for bit in bits:
        if bit == 1:
            count += 1
    return '1' if count % 2 == 0 else '0'


def generate_hamming_parity(bits, r, m):
    map = {}

    # Guardar en un mapa los valores de las potencias de 2 de cada m
    for i in range(len(m)):
        powers = sum_powers_of_2(m[i])
        map[m[i]] = powers

    parities = []
    for i in range(len(r)):
        parities.append(calc_parity(map, r[i], bits))

    new_parity_bits = ""
    for parity in parities:
        new_parity_bits += even_parity(parity)

    return new_parity_bits


def calc_parity(map, bit, plot):
    powers = []
    for key, values in map.items():
        if bit in values:
            powers.append(key)

    parity_check = []
    for power in powers:
        parity_check.append(int(plot[power - 1]))

    return parity_check


def sum_powers_of_2(v):
    copy_v = v
    powers = []
    while copy_v > 0:
        power = 1
        acu = 0
        while power <= copy_v:
            power *= 2
            acu += 1
        power //= 2
        copy_v -= power
        powers.append(acu - 1)

    powers_array = [2 ** power for power in powers]
    return powers_array


def find_error(b1, b2):
    b3 = ""
    for i in range(len(b1)):
        if b1[i] == b2[i]:
            b3 += '0'
        else:
            b3 += '1'

    b3 = b3[::-1]

    error = from_binary_to_decimal(b3)
    if error == 0:
        #print("No hay error")
        return 0
    else:
        #print("Error en la posicion:", error)
        return error


def fix_error(bits, error):
    new_bits = ""
    for i in range(len(bits)):
        if i == error - 1:
            new_bits += '1' if bits[i] == '0' else '0'
        else:
            new_bits += bits[i]
    return new_bits


def original(bits, r):
    new_bits = ""
    for i in range(len(bits)):
        if (i + 1) not in r:
            new_bits += bits[i]
    return new_bits


def binary_to_ascii(bits):
    n = len(bits)
    ascii = ""
    for i in range(0, n, 8):
        ascii += chr(from_binary_to_decimal(bits[i:i + 8]))
    return ascii




def main(received_hamming_code):
    # Received Hamming code (You can receive this code from the transmitter)
    #
    #print("Received Hamming code:", received_hamming_code)

    m = len(received_hamming_code)
    r = calculate_parity_bits(m)

    received_parity_bits = ""
    for i in r:
        received_parity_bits += received_hamming_code[i - 1]

    m_pos = [i + 1 for i in range(m) if (i + 1) not in r]

    new_parity_bits = generate_hamming_parity(received_hamming_code, r, m_pos)
    car = ''
    # error points to the position of the error in the received Hamming code, not the amount of errors
    error = find_error(received_parity_bits, new_parity_bits)
    if error > 0:
        corrected_hamming_code = fix_error(received_hamming_code, error)
        #print("Corrected Hamming code:", corrected_hamming_code)
        corrected_message = original(corrected_hamming_code, r)
        #print("Corrected message:", corrected_message)
        ascii = binary_to_ascii(corrected_message)
        car = ascii
    else:
        original_message = original(received_hamming_code, r)
        #print("Mensaje:", original_message)
        ascii = binary_to_ascii(original_message)
        #print("ASCII:", ascii)
        car = ascii

    #print('Caracter: ', car)
    # devolver el caracter, si hubo error y la cantidad de bits de paridad
    return car, error, len(r)

    


def receive_Message(hamming):
    # binary_codes = string_data[1:-1].split(', ')
    error_acum = 0
    parity_acum = 0
    codes_list = hamming[1:-1].split(', ')
    chars = []
    for code in codes_list:
        #chars.append(main(str(code)))
        c, e, p = main(str(code))
        chars.append(c)
        if e > 0:
            #print('Error encontrado en el codigo: ', code)
            error_acum += 1
        parity_acum += p
            
    
    return ''.join(chars), error_acum, parity_acum

File: Hamming/java/HammingR.java
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.lang.Math;
import java.util.HashMap;
//import java.util.Scanner;


public class HammingR {
    // Function to calculate the number of parity bits required
    private static int[] calculateParityBits(int m) {
        // given a message of length m, calculate the number of parity bits required and their positions
        int[] positions;
        int acu = 0;
        int i = 0;
        while (acu < m) {
            acu += Math.pow(2, i);
            i++;
        }
        positions = new int[i];
        for (int j = 0; j < i; j++) {
            positions[j] = (int) Math.pow(2, j);
        }
        //System.out.println("Number of parity bits required: " + positions.length);
        return positions;
            
    }

    
    //public static void main(String[] args) 
    public static String caracter(String receivedHammingCode)
    {
        //Scanner input = new Scanner(System.in);
        // Received Hamming code (You can receive this code from the transmitter)
        //String receivedHammingCode = "00100001001";
        //receivedHammingCode = "01100001001"; // el segundo bit esta mal
        //receivedHammingCode = "00100101001"; // el sexto bit esta mal
        // el sexto y el segundo bit estan mal
        //receivedHammingCode = "00100111011";

        // input code
        //receivedHammingCode = input.nextLine();
        //input.close();

        //System.out.println("Received Hamming code: " + receivedHammingCode);
        int m = receivedHammingCode.length();
        int[] r = calculateParityBits(m);
        String receivedParityBits = "";
        //String r_String = Arrays.toString(r);
        //System.out.println(r_String);
    
        for (int i = 0; i < r.length; i++) {
            //System.out.println("r[" + r[i] + "] = " + receivedHammingCode.charAt(r[i] - 1));
            receivedParityBits += receivedHammingCode.charAt(r[i] - 1);
        }

        int[] m_pos = new int[m - r.length];
        int acu = 0;

        for (int i = 1; i <= receivedHammingCode.length(); i++) {
            if (isInArray(r, i) == false) {
                //receivedHammingCode = receivedHammingCode.substring(0, i - 1) + receivedHammingCode.substring(i);
                //System.out.println("i = " + i);
                m_pos[acu] = i;
                acu++;
            }
        }
        //System.out.println("Received parity bits: " + receivedParityBits);
        //System.out.println(Arrays.toString(m_pos));
        String new_parity_bits = generate_hamming_parity(receivedHammingCode, r, m_pos);
        //System.out.println("New parity bits: " + new_parity_bits);
        String letter = "";
        int error = findError(receivedParityBits, new_parity_bits);
        if (error > 0){
            String correctedHammingCode = fixError(receivedHammingCode, error);
            //System.out.println("Corrected Hamming code: " + correctedHammingCode);
            String correctedMessage = original(correctedHammingCode, r);
            //System.out.println("Corrected message: " + correctedMessage);
            letter = correctedMessage;
        }
        else {
            String originalMessage = original(receivedHammingCode, r);
            //System.out.println("Mensaje: " +  originalMessage);
            letter = originalMessage;
        }
        letter = fromBinaryToAscii(letter);
        return letter;
        
    }

    public static String fromBinaryToAscii(String b){
        String ascii = "";
        int decimal_value = Integer.parseInt(b, 2);
        char c = (char) decimal_value;
        ascii += c;
        return ascii;
    }


    public static void main(String[] args) {
        // ingresar el arreglo de códigos hamming
        String[] helloWorld = {"01100001100", "01111000001", "01101101010", "10111011011", "0010000000", "00110101011", "11111010111", "11101101100", "00111011100", "01101101011", "0010000000", "0011100111"};
        String mensaje = "";
        for (int i = 0; i < helloWorld.length; i++) {
            mensaje += caracter(helloWorld[i]);
        }
        System.out.println("El mensaje es: " + mensaje);
        String[] helloWorld2 = {"01100001100", "10111001001", "01101101010", "00101010011", "1110000000", "00110100011", "10101011111", "01101100101", "00101011100", "10101100011", "1110000000", "1011100011"};
        String mensaje2 = "";
        for (int i = 0; i < helloWorld2.length; i++) {
            mensaje2 += caracter(helloWorld2[i]);
        }
        System.out.println("El mensaje es: " + mensaje2);
    }


    public static int fromBinaryToDecimal(String binary) {
        int decimal = 0;
        int n = binary.length();
        for (int i = 0; i < n; i++) {
            decimal += Integer.parseInt(binary.charAt(i) + "") * Math.pow(2, n - i - 1);
        }
        return decimal;
    }

    public static boolean isInArray(int[] array, int value) {
        for (int i = 0; i < array.length; i++) {
            if (value == array[i]) {
                return true;
            }
        }
        return false;
    }

    public static char even_parity(int[] bits){
        int count = 0;
        for (int i = 0; i < bits.length; i++) {
            if (bits[i] == 1) {
                count++;
            }
        }
        if (count % 2 == 0) {
            return '0';
        }
        else {
            return '1';
        }
    }
        
    public static char odd_parity(int[] bits){
        int count = 0;
        for (int i = 0; i < bits.length; i++) {
            if (bits[i] == 1) {
                count++;
            }
        }
        if (count % 2 == 0) {
            return '1';
        }
        else {
            return '0';
        }
    }
    
    public static String generate_hamming_parity(String bits, int[] r, int[] m ){
        HashMap<Integer, Integer[]> map = new HashMap<Integer, Integer[]>();

        // Guardar en un mapa los valores de las potencias de 2 de cada m
        for (int i = 0; i < m.length; i++){
            int[] powers = sum_powers_of_2(m[i]);
            //System.out.println("Powers of " + m[i] + ": " + Arrays.toString(powers));
            Integer[] powers_obj = new Integer[powers.length];
            for (int j = 0; j < powers.length; j++) {
                powers_obj[j] = powers[j];
            }
            map.put(m[i], powers_obj);
        }
        // print the hashmap

        /*
         * for (Map.Entry<Integer, Integer[]> entry : map.entrySet()) {
            Integer[] values = entry.getValue();
            System.out.println(entry.getKey() + " " + Arrays.toString(values));
        }
         */

        //System.out.println("Bits: " + Arrays.toString(r));

        List <int[]> parities = new ArrayList<int[]>();
        for (int i = 0; i < r.length; i++) {
            //int parity = calc_parity(map, r[i], bits);
            //System.out.println("Parity: " + parity);
            parities.add(calc_parity(map, r[i], bits));
        }

        /*for (int i = 0; i < parities.size(); i++) {
            System.out.println("Parity " + (i + 1) + ": " + Arrays.toString(parities.get(i)));
        }*/

        String new_parity_bits = "";
        for (int i = 0; i < parities.size(); i++) {
            new_parity_bits += even_parity(parities.get(i));
        }
        
        return new_parity_bits;
    }

    public static int[] calc_parity(HashMap<Integer, Integer[]> map, int bit, String plot){

        List <Integer> powers = new ArrayList<Integer>();
        for (Map.Entry<Integer, Integer[]> entry : map.entrySet()) {
            Integer[] values = entry.getValue();
            //System.out.println(" - " + entry.getKey() + " -- " + Arrays.toString(values));
            int[] values_int = new int[values.length];
            for (int i = 0; i < values.length; i++) {
                values_int[i] = values[i];
            }
            //System.out.println(" - " + entry.getKey() + " -- " + Arrays.toString(values_int));
            //System.out.println("b = " + bit);
            if (isInArray(values_int, bit)) {
                powers.add(entry.getKey());
            }
        }
        
        List <Integer> parity_check = new ArrayList<Integer>();

        
        for (int i = 0; i < powers.size(); i++) {
            //System.out.println(" - " + powers.get(i) + " -- " + plot.charAt(powers.get(i) - 1));
            parity_check.add(Integer.parseInt(plot.charAt(powers.get(i) - 1) + ""));
        }

        //System.out.println(" - " + bit + " -- " + parity_check);
        int[] parity_check_array = new int[parity_check.size()];
        for (int i = 0; i < parity_check.size(); i++) {
            parity_check_array[i] = parity_check.get(i);
        }
        return parity_check_array;
    }

    public static int[] sum_powers_of_2(int v){

        int copy_v = v;
        List <Integer> powers = new ArrayList<Integer>();
        while (copy_v > 0) {
            int power = 1;
            int acu = 0;
            while (power <= copy_v) {
                power *= 2;
                acu++;
            }
            power /= 2;
            copy_v -= power;
            powers.add(acu - 1);
        }

        //System.out.println(" - " + v + " -- " + powers);

        int[] powers_array = new int[powers.size()];
        for (int i = 0; i < powers.size(); i++) {
            double potencia = Math.pow(2, powers.get(i));
            powers_array[i] = (int) potencia;
        }
        return powers_array;
    }

    public static int findError(String b1, String b2){
        String b3 = "";
        for (int i = 0; i < b1.length(); i++) {
            if (b1.charAt(i) == b2.charAt(i)) {
                b3 += '0';
            }
            else {
                b3 += '1';
            }
        }

        //System.out.println("b3: " + b3);
        b3 = reverseStringUsingLoop(b3);
        //System.out.println("b3: " + b3);
        
        int error = fromBinaryToDecimal(b3);
        //System.out.println("Error: " + error);
        if (error == 0){
            //System.out.println("No hay error");
            return 0;
        }
        else {
            System.out.println("Error en la posicion: " + error);
            return error;
        }
    }

    public static String reverseStringUsingLoop(String input) {
        StringBuilder reversed = new StringBuilder();

        for (int i = input.length() - 1; i >= 0; i--) {
            reversed.append(input.charAt(i));
        }

        return reversed.toString();
    }

    public static String fixError(String bits, int error){
        String new_bits = "";
        for (int i = 0; i < bits.length(); i++) {
            if (i == error - 1) {
                if (bits.charAt(i) == '0') {
                    new_bits += '1';
                }
                else {
                    new_bits += '0';
                }
            }
            else {
                new_bits += bits.charAt(i);
            }
        }
        return new_bits;
    }

    public static String original(String bits, int[] r){
        // remove parity bits
        String new_bits = "";
        
        for (int i = 0; i < bits.length(); i++) {
            if (!isInArray(r, i + 1)) {
                new_bits += bits.charAt(i);
            }
        }
        return new_bits;

    }
}
